components/constellation/constellation.rs

```diff
@@ -1396,20 +1402,65 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         };
 
         // Create the new pipeline, attached to the parent and push to pending frames
-        self.new_pipeline(load_info.new_pipeline_id,
-                          load_info.frame_id,
-                          Some((load_info.parent_pipeline_id, load_info.frame_type)),
+        self.new_pipeline(load_info.info.new_pipeline_id,
+                          load_info.info.frame_id,
+                          Some((load_info.info.parent_pipeline_id, load_info.info.frame_type)),
                           window_size,
                           load_data,
                           load_info.sandbox,
                           is_private);
 
         self.pending_frames.push(FrameChange {
-            frame_id: load_info.frame_id,
+            frame_id: load_info.info.frame_id,
             old_pipeline_id: load_info.old_pipeline_id,
-            new_pipeline_id: load_info.new_pipeline_id,
+            new_pipeline_id: load_info.info.new_pipeline_id,
+            document_ready: false,
+            replace: load_info.info.replace,
+        });
+    }
+
+    fn handle_script_loaded_about_blank_in_iframe_msg(&mut self,
+                                                      load_info: IFrameLoadInfo,
+                                                      layout_sender: IpcSender<LayoutControlMsg>) {
+        let IFrameLoadInfo {
+            parent_pipeline_id,
+            new_pipeline_id,
+            frame_type,
+            replace,
+            frame_id,
+            is_private,
+        } = load_info;
+
+        let pipeline = {
+            let parent_pipeline = match self.pipelines.get(&parent_pipeline_id) {
+                Some(parent_pipeline) => parent_pipeline,
+                None => return warn!("Script loaded url in closed iframe {}.", parent_pipeline_id),
+            };
+
+            let script_sender = parent_pipeline.script_chan.clone();
+
+            let url = ServoUrl::parse("about:blank").expect("infallible");
+            Pipeline::spawned(new_pipeline_id,
```

Comment:
It does not seem like the new function shares any useful fraction of its code with `new_pipeline`.

Currently, `new_frame` is called in response to `ConstellationMsg::ActivateDocument`; it seems like a significant amount of work to call it earlier.